R/lump.R

```diff
@@ -56,46 +58,33 @@
 #'
 fct_lump <- function(f, n, prop, w = NULL, other_level = "Other",
                      ties.method = c("min", "average", "first", "last", "random", "max")) {
-  f <- check_factor(f)
-  w <- check_weights(w, length(f))
-  ties.method <- match.arg(ties.method)
 
+  ties.method <- match.arg(ties.method)
+  calcs <- calc_levels(f, w)
+  f <- calcs$f
+  count <- calcs$count
+  total <- calcs$total
   levels <- levels(f)
-  if (is.null(w)) {
-    count <- as.vector(table(f))
-    total <- length(f)
-  } else {
-    count <- as.vector(tapply(w, f, FUN = sum))
-    total <- sum(w)
-  }
 
   if (!xor(missing(n), missing(prop))) {
     new_levels <- ifelse(!in_smallest(count), levels, other_level)
   } else if (!missing(n)) {
-    if (n < 0) {
-      rank <- rank(count, ties = ties.method)
-      n <- -n
-    } else {
-      rank <- rank(-count, ties = ties.method)
-    }
+    check_rank <- .fct_lump_n(n, count, levels, other_level, ties.method)
 
-    if (sum(rank > n) <= 1) {
+    if (sum(check_rank$rank > n) <= 1) {
       # No lumping needed
       return(f)
+    } else {
+      new_levels <- check_rank$new_levels
     }
 
-    new_levels <- ifelse(rank <= n, levels, other_level)
   } else if (!missing(prop)) {
-    prop_n <- count / total
-    if (prop < 0) {
-      new_levels <- ifelse(prop_n <= -prop, levels, other_level)
+    check_prop <- .fct_lump_prop(prop, count, total, levels, other_level)
```

Comment:
Similarly, why doesn't this call `fct_lump_prop()`? I think `fct_lump()` should just become a wrapper that automatically picks which lower level function you want.

Reply:
Done.